### ISSUE TITLE
Complete MLX Pipeline: Training → Fine-Tuning → Inference

### ISSUE THREAD

**Issue by akaszubski** [akaszubski/realign#47]
## Overview
Implement end-to-end MLX pipeline integrating all three components for a complete machine learning workflow.

## Pipeline Architecture

```
┌─────────────────────────────────────────────────────────────────┐
│                     REALIGN MLX PIPELINE                         │
└─────────────────────────────────────────────────────────────────┘

Phase 1: FINE-TUNING (MLX-LM)
├─> Load base model (4-bit quantized)
├─> LoRA/QLoRA fine-tuning on realign dataset
├─> Fuse adapters
└─> Save fine-tuned model
    │
    ▼
Phase 2: DISTRIBUTED TRAINING (Optional - MLX-DISTRIBUTED-TRAINING)
├─> Load fine-tuned model
├─> Continue training across multiple devices
├─> Gradient synchronization via MPI
└─> Save checkpoints
    │
    ▼
Phase 3: PRODUCTION INFERENCE (MLX_PARALLM)
├─> Load final trained model
├─> Batch inference (500+ prompts)
└─> High-throughput production deployment
```

## Dependencies
- Issue #44: MLX-LM Integration
- Issue #45: MLX-DISTRIBUTED-TRAINING Integration  
- Issue #46: MLX_PARALLM Integration

### 1. Pipeline Orchestration
- [ ] Create unified configuration system
- [ ] Implement phase transitions (fine-tuning → training → inference)
- [ ] Add model checkpointing between phases
- [ ] Setup automatic model versioning

### 2. Data Management
- [ ] Unified data format across all phases
- [ ] Data validation and preprocessing
- [ ] Train/validation/test splits
- [ ] Dataset versioning and tracking

### 3. Training Workflow
- [ ] Implement Phase 1: Fine-tuning script
- [ ] Implement Phase 2: Distributed training script (optional)
- [ ] Add automatic hyperparameter tuning
- [ ] Setup experiment tracking (W&B/SwanLab)

### 4. Deployment Workflow
- [ ] Implement Phase 3: Production inference API
- [ ] Add model serving endpoint
- [ ] Setup request batching and queuing
- [ ] Add monitoring and metrics

### 5. Integration Testing
- [ ] End-to-end pipeline test
- [ ] Verify model quality across phases
- [ ] Performance benchmarking
- [ ] Resource utilization monitoring

### 6. Documentation
- [ ] Pipeline architecture documentation
- [ ] Usage guide for each phase
- [ ] Troubleshooting guide
- [ ] Performance optimization guide

## Example End-to-End Workflow

```python
# ============================================================================
# PHASE 1: FINE-TUNING WITH MLX-LM
# ============================================================================
import subprocess

# Fine-tune base model with LoRA
subprocess.run([
    "mlx_lm.lora",
    "--model", "mlx-community/Llama-3.2-3B-4bit",
    "--train",
    "--data", "./realign_dataset",
    "--iters", "2000",
    "--batch-size", "4",
    "--grad-accumulation-steps", "4",
    "--adapter-path", "./adapters/realign_v1"
])

# Fuse adapters
subprocess.run([
    "mlx_lm.fuse",
    "--model", "mlx-community/Llama-3.2-3B-4bit",
    "--adapter-path", "./adapters/realign_v1",
    "--output-path", "./models/realign_finetuned"
])

# ============================================================================
# PHASE 2 (OPTIONAL): DISTRIBUTED TRAINING
# ============================================================================
from src.training.distributed_trainer import DistributedTrainer, TrainingConfig
from src.models.unified_model import UnifiedModel

# Load fine-tuned model
model = UnifiedModel.from_pretrained("./models/realign_finetuned")

# Continue training across multiple devices
config = TrainingConfig(
    batch_size=16,
    gradient_accumulation_steps=4,
    max_memory_gb=32,
    mixed_precision=True
)

trainer = DistributedTrainer(config=config, model=model)
trainer.train(dataset="large_realign_dataset", num_epochs=3)

# ============================================================================
# PHASE 3: PRODUCTION INFERENCE WITH MLX_PARALLM
# ============================================================================
from mlx_parallm.utils import load, batch_generate

# Load final trained model
model, tokenizer = load("./models/realign_final")

# Production batch inference
user_queries = load_queries()  # Load user queries

responses = batch_generate(
    model,
    tokenizer,
    prompts=user_queries,
    max_tokens=256,
    temp=0.0,  # Deterministic for production
    format_prompts=True,
    verbose=True
)

# Deploy responses
save_responses(responses)
```

## Performance Targets
- **Fine-tuning**: Complete in <2 hours on 32GB device
- **Distributed training**: 38%+ speedup with 2 devices
- **Inference**: 1000+ tokens/sec throughput
- **Latency**: <100ms per query in production

### Minimum (Single Device)
- Apple Silicon M1/M2/M3
- 32GB RAM
- Fine-tuning + Inference only

### Recommended (Distributed)
- 2+ Apple Silicon devices
- 64GB+ RAM per device
- 10Gbps network or Thunderbolt 4
- Full pipeline support

### Optimal (Production)
- 3+ devices (training + inference cluster)
- 128GB+ RAM for float16 models
- High-speed network
- Maximum throughput

## Success Criteria
- [ ] Complete end-to-end run from data → deployment
- [ ] Model quality meets benchmarks
- [ ] Inference throughput >1000 tok/s
- [ ] Pipeline is reproducible and documented
- [ ] All components integrate seamlessly

## Documentation References
- MLX-LM: `~/Documents/GitHub/distributed-mlx/mlx-lm/`
- Distributed Training: `~/Documents/GitHub/distributed-mlx/mlx-disitrubted-training/`
- Batch Inference: `~/Documents/GitHub/distributed-mlx/mlx_parallm/`

## Labels
epic, mlx, pipeline, integration, priority:critical

**Comment by akaszubski** [comment]
## 🔄 Related to Issue #48

This end-to-end pipeline issue is related to **#48: Delegate all training methods to mlx-lm-lora**

### Updated Architecture (Based on #48)

```
┌─────────────────────────────────────────────────────────────┐
│              REALIGN MLX PIPELINE v3.2.0+                   │
└─────────────────────────────────────────────────────────────┘

Phase 1: TRAINING (mlx-lm-lora via #48)
├─> LoRA/QLoRA/DoRA fine-tuning
├─> DPO/ORPO/GRPO/CPO preference optimization
├─> Full fine-tuning
└─> Distributed training (built-in mlx.distributed)
    │
    ▼
Phase 2: INFERENCE (This Issue)
├─> MLX_PARALLM for batch inference (#46)
├─> mlx-sharding for distributed inference (experimental)
└─> OpenAI-compatible API serving
```

### Key Changes from #48

**Training** (Issue #48):
- ✅ Uses mlx-lm-lora (not mlx-lm alone)
- ✅ Distributed via built-in mlx.distributed (not external package)
- ✅ All advanced methods (DPO, ORPO, GRPO, CPO)

**This Issue** should focus on:
- Inference optimization (MLX_PARALLM)
- Serving infrastructure
- API integration
- Evaluation workflows

### Recommendation

Update this issue to:
1. Reference #48 for training phase
2. Focus on inference/serving phases
3. Coordinate with #46 for batch inference
4. Use mlx-lm-lora outputs as inputs

See #48 comments and ADR-002 for distributed training details.
